File: languages/ruby/lib/oso/polar/predicate.rb
# frozen_string_literal: true

class Oso
  class Polar
    # Polar predicate.
    class Predicate
      attr_reader :name, :args

      # @param name [String]
      # @param args [Array<Object>]
      def initialize(name, args:)
        @name = name
        @args = args
      end

      # @param other [Predicate]
      # @return [Boolean]
      def ==(other)
        name == other.name && args == other.args
      end

      # @see #==
      alias eql? ==
    end
  end
end

File: languages/python/oso/polar/predicate.py
from typing import Any, Sequence


class Predicate:
    """Represent a predicate in Polar (`name(args, ...)`)."""

    name: str
    args: Sequence[Any]

    def __init__(self, name: str, args: Sequence[Any]) -> None:
        self.name = name
        self.args = args

    def __str__(self) -> str:
        return f'{self.name}({", ".join(self.args)})'

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Predicate):
            return False
        return (
            self.name == other.name
            and len(self.args) == len(other.args)
            and all(x == y for x, y in zip(self.args, other.args))
        )
